File: ts/src/utils/preprocessing.ts
/**
 * Data preprocessing utilities for time series
 *
 * @module utils/preprocessing
 */

/**
 * Normalize data to [0, 1] range (min-max scaling)
 *
 * @param data - Input data
 * @returns Object with normalized data, min, and max values
 *
 * @example
 * ```typescript
 * import { normalize, denormalize } from 'rustful-ts';
 *
 * const data = [10, 20, 30, 40, 50];
 * const { normalized, min, max } = normalize(data);
 * // normalized: [0, 0.25, 0.5, 0.75, 1]
 *
 * // Later, convert back
 * const original = denormalize(normalized, min, max);
 * ```
 */
export function normalize(data: number[]): {
  normalized: number[];
  min: number;
  max: number;
} {
  if (data.length === 0) {
    return { normalized: [], min: 0, max: 1 };
  }

  const min = Math.min(...data);
  const max = Math.max(...data);
  const range = max - min;

  if (Math.abs(range) < 1e-10) {
    return { normalized: data.map(() => 0.5), min, max };
  }

  const normalized = data.map((x) => (x - min) / range);
  return { normalized, min, max };
}

/**
 * Denormalize data from [0, 1] range
 */
export function denormalize(data: number[], min: number, max: number): number[] {
  const range = max - min;
  return data.map((x) => x * range + min);
}

/**
 * Standardize data to zero mean and unit variance (z-score)
 *
 * @param data - Input data
 * @returns Object with standardized data, mean, and standard deviation
 */
export function standardize(data: number[]): {
  standardized: number[];
  mean: number;
  stdDev: number;
} {
  if (data.length === 0) {
    return { standardized: [], mean: 0, stdDev: 1 };
  }

  const n = data.length;
  const mean = data.reduce((a, b) => a + b, 0) / n;
  const variance = data.reduce((acc, x) => acc + (x - mean) ** 2, 0) / n;
  const stdDev = Math.sqrt(variance);

  if (stdDev < 1e-10) {
    return { standardized: data.map(() => 0), mean, stdDev: 1 };
  }

  const standardized = data.map((x) => (x - mean) / stdDev);
  return { standardized, mean, stdDev };
}

/**
 * Destandardize data
 */
export function destandardize(
  data: number[],
  mean: number,
  stdDev: number
): number[] {
  return data.map((x) => x * stdDev + mean);
}

/**
 * Apply log transformation (log1p for handling zeros)
 */
export function logTransform(data: number[]): number[] {
  return data.map((x) => Math.log1p(x));
}

/**
 * Inverse log transformation
 */
export function inverseLogTransform(data: number[]): number[] {
  return data.map((x) => Math.expm1(x));
}

/**
 * Compute first-order differences
 *
 * @param data - Input data
 * @param order - Number of times to difference (default: 1)
 * @returns Differenced series (length reduced by `order`)
 *
 * @example
 * ```typescript
 * import { difference } from 'rustful-ts';
 *
 * const data = [1, 3, 6, 10];
 * console.log(difference(data, 1)); // [2, 3, 4]
 * console.log(difference(data, 2)); // [1, 1]
 * ```
 */
export function difference(data: number[], order: number = 1): number[] {
  let result = [...data];
  for (let i = 0; i < order; i++) {
    if (result.length <= 1) {
      return [];
    }
    const differenced: number[] = [];
    for (let j = 1; j < result.length; j++) {
      differenced.push(result[j] - result[j - 1]);
    }
    result = differenced;
  }
  return result;
}

/**
 * Compute seasonal differences
 *
 * @param data - Input data
 * @param period - Seasonal period
 * @returns Seasonally differenced series
 */
export function seasonalDifference(data: number[], period: number): number[] {
  if (data.length <= period) {
    return [];
  }

  const result: number[] = [];
  for (let i = period; i < data.length; i++) {
    result.push(data[i] - data[i - period]);
  }
  return result;
}

/**
 * Remove NaN and infinite values
 */
export function cleanData(data: number[]): number[] {
  return data.filter((x) => Number.isFinite(x));
}

/**
 * Linear interpolation for missing values (NaN)
 */
export function interpolateLinear(data: number[]): number[] {
  const result = [...data];
  const n = result.length;

  for (let i = 0; i < n; i++) {
    if (Number.isNaN(result[i])) {
      // Find previous valid value
      let prevIdx = -1;
      for (let j = i - 1; j >= 0; j--) {
        if (!Number.isNaN(result[j])) {
          prevIdx = j;
          break;
        }
      }

      // Find next valid value
      let nextIdx = -1;
      for (let j = i + 1; j < n; j++) {
        if (!Number.isNaN(result[j])) {
          nextIdx = j;
          break;
        }
      }

      if (prevIdx >= 0 && nextIdx >= 0) {
        const ratio = (i - prevIdx) / (nextIdx - prevIdx);
        result[i] = result[prevIdx] + ratio * (result[nextIdx] - result[prevIdx]);
      } else if (prevIdx >= 0) {
        result[i] = result[prevIdx];
      } else if (nextIdx >= 0) {
        result[i] = result[nextIdx];
      } else {
        result[i] = 0;
      }
    }
  }

  return result;
}

/**
 * Detect outliers using IQR method
 *
 * @param data - Input data
 * @param multiplier - IQR multiplier (default: 1.5)
 * @returns Indices of outlier points
 */
export function detectOutliersIQR(
  data: number[],
  multiplier: number = 1.5
): number[] {
  const sorted = [...data].filter(Number.isFinite).sort((a, b) => a - b);

  if (sorted.length < 4) {
    return [];
  }

  const q1 = sorted[Math.floor(sorted.length / 4)];
  const q3 = sorted[Math.floor((3 * sorted.length) / 4)];
  const iqr = q3 - q1;

  const lowerBound = q1 - multiplier * iqr;
  const upperBound = q3 + multiplier * iqr;

  return data
    .map((x, i) => (x < lowerBound || x > upperBound ? i : -1))
    .filter((i) => i >= 0);
}

/**
 * Create lagged features for machine learning
 *
 * @param data - Input time series
 * @param lags - Number of lag features to create
 * @returns Array of feature vectors (each row is [lag1, lag2, ..., lagN, target])
 */
export function createLagFeatures(
  data: number[],
  lags: number
): { features: number[][]; target: number[] } {
  if (data.length <= lags) {
    return { features: [], target: [] };
  }

  const features: number[][] = [];
  const target: number[] = [];

  for (let i = lags; i < data.length; i++) {
    const row: number[] = [];
    for (let j = lags; j >= 1; j--) {
      row.push(data[i - j]);
    }
    features.push(row);
    target.push(data[i]);
  }

  return { features, target };
}

/**
 * Split data into train and test sets (respecting time order)
 *
 * @param data - Input data
 * @param testRatio - Fraction of data for testing (0-1)
 * @returns Object with train and test arrays
 */
export function trainTestSplit(
  data: number[],
  testRatio: number = 0.2
): { train: number[]; test: number[] } {
  const ratio = Math.max(0.1, Math.min(0.9, testRatio));
  const splitIdx = Math.floor((1 - ratio) * data.length);
  const boundedSplitIdx = Math.max(1, Math.min(data.length - 1, splitIdx));

  return {
    train: data.slice(0, boundedSplitIdx),
    test: data.slice(boundedSplitIdx),
  };
}

File: crates/rustful-core/src/utils/preprocessing.rs
//! Data preprocessing utilities for time series
//!
//! Functions for cleaning, transforming, and preparing time series data.

/// Normalize data to [0, 1] range (min-max scaling)
pub fn normalize(data: &[f64]) -> (Vec<f64>, f64, f64) {
    if data.is_empty() {
        return (Vec::new(), 0.0, 1.0);
    }

    let min = data.iter().cloned().fold(f64::INFINITY, f64::min);
    let max = data.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
    let range = max - min;

    if range.abs() < 1e-10 {
        return (vec![0.5; data.len()], min, max);
    }

    let normalized: Vec<f64> = data.iter().map(|x| (x - min) / range).collect();

    (normalized, min, max)
}

/// Denormalize data from [0, 1] range
pub fn denormalize(data: &[f64], min: f64, max: f64) -> Vec<f64> {
    let range = max - min;
    data.iter().map(|x| x * range + min).collect()
}

/// Standardize data to zero mean and unit variance (z-score)
pub fn standardize(data: &[f64]) -> (Vec<f64>, f64, f64) {
    if data.is_empty() {
        return (Vec::new(), 0.0, 1.0);
    }

    let n = data.len() as f64;
    let mean = data.iter().sum::<f64>() / n;
    let variance = data.iter().map(|x| (x - mean).powi(2)).sum::<f64>() / n;
    let std_dev = variance.sqrt();

    if std_dev < 1e-10 {
        return (vec![0.0; data.len()], mean, 1.0);
    }

    let standardized: Vec<f64> = data.iter().map(|x| (x - mean) / std_dev).collect();

    (standardized, mean, std_dev)
}

/// Destandardize data
pub fn destandardize(data: &[f64], mean: f64, std_dev: f64) -> Vec<f64> {
    data.iter().map(|x| x * std_dev + mean).collect()
}

/// Apply log transformation (log1p for handling zeros)
pub fn log_transform(data: &[f64]) -> Vec<f64> {
    data.iter().map(|x| (x + 1.0).ln()).collect()
}

/// Inverse log transformation
pub fn inverse_log_transform(data: &[f64]) -> Vec<f64> {
    data.iter().map(|x| x.exp() - 1.0).collect()
}

/// Apply Box-Cox transformation
///
/// # Arguments
///
/// * `data` - Input data (must be positive)
/// * `lambda` - Transformation parameter (0 = log transform)
pub fn box_cox(data: &[f64], lambda: f64) -> Vec<f64> {
    if lambda.abs() < 1e-10 {
        data.iter().map(|x| x.ln()).collect()
    } else {
        data.iter().map(|x| (x.powf(lambda) - 1.0) / lambda).collect()
    }
}

/// Inverse Box-Cox transformation
pub fn inverse_box_cox(data: &[f64], lambda: f64) -> Vec<f64> {
    if lambda.abs() < 1e-10 {
        data.iter().map(|x| x.exp()).collect()
    } else {
        data.iter()
            .map(|x| (x * lambda + 1.0).powf(1.0 / lambda))
            .collect()
    }
}

/// Remove NaN and infinite values
pub fn clean_data(data: &[f64]) -> Vec<f64> {
    data.iter()
        .filter(|x| x.is_finite())
        .cloned()
        .collect()
}

/// Interpolate missing values (marked as NaN)
pub fn interpolate_linear(data: &[f64]) -> Vec<f64> {
    let mut result = data.to_vec();
    let n = result.len();

    for i in 0..n {
        if result[i].is_nan() {
            // Find previous valid value
            let prev_idx = (0..i).rev().find(|&j| !result[j].is_nan());
            // Find next valid value
            let next_idx = ((i + 1)..n).find(|&j| !result[j].is_nan());

            result[i] = match (prev_idx, next_idx) {
                (Some(p), Some(n_idx)) => {
                    let ratio = (i - p) as f64 / (n_idx - p) as f64;
                    result[p] + ratio * (result[n_idx] - result[p])
                }
                (Some(p), None) => result[p],
                (None, Some(n_idx)) => result[n_idx],
                (None, None) => 0.0,
            };
        }
    }

    result
}

/// Detect outliers using IQR method
///
/// Returns indices of outlier points
pub fn detect_outliers_iqr(data: &[f64], multiplier: f64) -> Vec<usize> {
    let mut sorted: Vec<f64> = data.iter().filter(|x| x.is_finite()).cloned().collect();
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap());

    if sorted.len() < 4 {
        return Vec::new();
    }

    let q1 = sorted[sorted.len() / 4];
    let q3 = sorted[3 * sorted.len() / 4];
    let iqr = q3 - q1;

    let lower_bound = q1 - multiplier * iqr;
    let upper_bound = q3 + multiplier * iqr;

    data.iter()
        .enumerate()
        .filter(|(_, x)| **x < lower_bound || **x > upper_bound)
        .map(|(i, _)| i)
        .collect()
}

/// Compute first-order differences
pub fn difference(data: &[f64], order: usize) -> Vec<f64> {
    let mut result = data.to_vec();
    for _ in 0..order {
        if result.len() <= 1 {
            return Vec::new();
        }
        result = result.windows(2).map(|w| w[1] - w[0]).collect();
    }
    result
}

/// Compute seasonal differences
pub fn seasonal_difference(data: &[f64], period: usize) -> Vec<f64> {
    if data.len() <= period {
        return Vec::new();
    }

    data.iter()
        .skip(period)
        .zip(data.iter())
        .map(|(curr, prev)| curr - prev)
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_normalize() {
        let data = vec![0.0, 5.0, 10.0];
        let (normalized, min, max) = normalize(&data);

        assert!((min - 0.0).abs() < 1e-10);
        assert!((max - 10.0).abs() < 1e-10);
        assert!((normalized[0] - 0.0).abs() < 1e-10);
        assert!((normalized[1] - 0.5).abs() < 1e-10);
        assert!((normalized[2] - 1.0).abs() < 1e-10);
    }

    #[test]
    fn test_standardize() {
        let data = vec![1.0, 2.0, 3.0, 4.0, 5.0];
        let (standardized, mean, std_dev) = standardize(&data);

        assert!((mean - 3.0).abs() < 1e-10);
        assert!((standardized.iter().sum::<f64>()).abs() < 1e-10);

        let recovered = destandardize(&standardized, mean, std_dev);
        for (orig, rec) in data.iter().zip(recovered.iter()) {
            assert!((orig - rec).abs() < 1e-10);
        }
    }

    #[test]
    fn test_difference() {
        let data = vec![1.0, 3.0, 6.0, 10.0];
        let diff = difference(&data, 1);
        assert_eq!(diff, vec![2.0, 3.0, 4.0]);

        let diff2 = difference(&data, 2);
        assert_eq!(diff2, vec![1.0, 1.0]);
    }
}
